### PR TITLE
Make coroutine function return type more specific

### LINKED ISSUE #3569: Coroutine functions should return Coroutine, not Awaitable
```python
async def coro_func() -> None:
    pass
c = coro_func()
reveal_type(c)  # prints typing.Awaitable[builtins.None]
```

`c` above has all fields of `typing.Coroutine`, you can `send` and `throw` on it, but the type doesn't reflect that.

### LINKED ISSUE #4460: Question regarding Coroutine vs Awaitable
Given this code:

```python
from typing import *
from collections import deque

Task = Coroutine[Any, None, Any]
tasks: Deque[Task] = deque()


async def foo() -> None:
    print('foo!')


def add_task(task: Task) -> None:
    tasks.append(task)


def run() -> None:
    task = tasks.popleft()
    try:
        task.send(None)
    except StopIteration:
        pass


add_task(foo())
run()
```

Versions:

```
$ mypy --version
mypy 0.560-51e044c4ecf2a52fc6c41ee63019723e0d3061e1-dirty
$ python --version
Python 3.6.4
```

Running `mypy --strict script.py` I get `script.py:24: error: Argument 1 to "add_task" has incompatible type "Awaitable[None]"; expected "Coroutine[Any, None, Any]"`, but changing `Task` to `Awaitable[None]` gives me `script.py:19: error: "Awaitable[None]" has no attribute "send"`. The code runs fine though.

What would be the proper type hints in this scenario?

### PR DESCRIPTION
**Pull request by AndreLouisCaron**
See the commit messages for details and rationale.  I tried to "tell a story" with them :-)

Fixes #3569.
Fixes #4460.

Special thanks to @ilevkivskyi and @gvanrossum for their time and their infinite patience with all my questions :-)

### CODE REVIEW

File: mypy/checker.py [hunk 1]
```diff
@@ -578,6 +578,13 @@ def get_generator_receive_type(self, return_type: Type, is_coroutine: bool) -> T
             # values.  IOW, tc is None.
             return NoneTyp()
 
+    def get_coroutine_return_type(self, return_type: Type) -> Type:
+        if isinstance(return_type, AnyType):
+            return AnyType(TypeOfAny.from_another_any, source_any=return_type)
+        assert isinstance(return_type, Instance), "Should only be called on coroutine functions!"
```
[the remark this answers is not in the document]
Comment (AndreLouisCaron) [reply mypy/checker.py]: Fixed in latest commit :-)

File: mypy/checker.py [hunk 2]
```diff
@@ -578,6 +578,13 @@ def get_generator_receive_type(self, return_type: Type, is_coroutine: bool) -> T
...
+        # return type is 3rd type specification in Coroutine!
```
[the remark this answers is not in the document]
Comment (AndreLouisCaron) [reply mypy/checker.py]: Fixed in latest commit :-)

File: mypy/checker.py [hunk 3]
```diff
@@ -878,7 +890,8 @@ def is_unannotated_any(t: Type) -> bool:
                 if is_unannotated_any(ret_type):
                     self.fail(messages.RETURN_TYPE_EXPECTED, fdef)
                 elif (fdef.is_coroutine and isinstance(ret_type, Instance) and
-                      is_unannotated_any(ret_type.args[0])):
+                      is_unannotated_any(ret_type.args[2])):
```
[the remark this answers is not in the document]
Comment (AndreLouisCaron) [reply mypy/checker.py]: Fixed in latest commit :-)

File: mypy/semanal.py [hunk 4]
```diff
@@ -431,9 +431,11 @@ def _visit_func_def(self, defn: FuncDef) -> None:
                     pass
                 else:
                     # A coroutine defined as `async def foo(...) -> T: ...`
-                    # has external return type `Awaitable[T]`.
-                    ret_type = self.named_type_or_none('typing.Awaitable', [defn.type.ret_type])
-                    assert ret_type is not None, "Internal error: typing.Awaitable not found"
+                    # has external return type `Coroutine[T, Any, Any]`.
```
[the remark this answers is not in the document]
Comment (AndreLouisCaron) [reply mypy/semanal.py]: Fixed in latest commit :-)

**Comment by AndreLouisCaron** [comment]
Thanks @ilevkivskyi for the quick feedback.  I pushed a new commit with all the requested changes :-)

File: mypy/checker.py [hunk 5]
```diff
@@ -578,6 +578,13 @@ def get_generator_receive_type(self, return_type: Type, is_coroutine: bool) -> T
...
             return NoneTyp()
 
+    def get_coroutine_return_type(self, return_type: Type) -> Type:
+        if isinstance(return_type, AnyType):
+            return AnyType(TypeOfAny.from_another_any, source_any=return_type)
+        assert isinstance(return_type, Instance), "Should only be called on coroutine functions."
+        # Note: return type is the 3rd type parameter of Coroutine.
+        return return_type.args[2]
+
     def get_generator_return_type(self, return_type: Type, is_coroutine: bool) -> Type:
```
Comment (JelleZijlstra) [inline mypy/checker.py]: Is the `is_coroutine` argument here still useful, or can we remove it now?

Comment (AndreLouisCaron) [reply mypy/checker.py]: I'm pretty sure we still need it to handle asynchronous generators.  In that case, they have both `is_generator=True` and `is_coroutine=True`.

**Comment by gvanrossum** [comment]
Thank you for fixing such a sophisticated problem! Vive les sprint. :-)

**Pull request merged**